interfaces/builtin/pkcs11.go

```diff
@@ -0,0 +1,166 @@
+  pkcs11:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+type pkcs11Interface struct{}
+
+// Name of the pkcs11 interface.
+func (iface *pkcs11Interface) Name() string {
+	return "pkcs11"
+}
+
+func (iface *pkcs11Interface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              pkcs11Summary,
+		ImplicitOnCore:       true,
+		ImplicitOnClassic:    true,
+		BaseDeclarationSlots: pkcs11BaseDeclarationSlots,
+		BaseDeclarationPlugs: pkcs11BaseDeclarationPlugs,
+	}
+}
+
+func (iface *pkcs11Interface) String() string {
+	return iface.Name()
+}
+
+const pkcs11PermanentSlotSecComp = `
+# Description: Allow operating as an p11-kit server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+
+# Needed by server upon client connect
+accept
+accept4
+`
+
+func (iface *pkcs11Interface) getSocketPath(slot *snap.SlotInfo) (string, error) {
+	socketAttr, isSet := slot.Attrs["pkcs11-socket"]
+	if !isSet {
+		return "", fmt.Errorf(`pkcs11 slot %s has missing "pkcs11-socket" attribute`, slot.Name)
+	}
+
+	socketPath, ok := socketAttr.(string)
+	if !ok {
+		return "", fmt.Errorf(`pkcs11: "pkcs11-socket" attribute must be a string, not %v`,
+			slot.Attrs["pkcs11-socket"])
+	}
+
+	// separate socket name and check socket path to start with /run/p11-kit/pkcs11-
+	if "/run/p11-kit" != filepath.Dir(socketPath) {
+		return "", fmt.Errorf("slot %q, a unix socket has to be in /run/p11-kit directory", slot.Name)
+	}
+
+	socketName := filepath.Base(socketPath)
+	if !strings.HasPrefix(socketName, "pkcs11-") {
+		return "", fmt.Errorf("slot %q, a unix socket name has to start with 'pkcs11-'", slot.Name)
+	}
+	return socketPath, nil
+}
+
+func (iface *pkcs11Interface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	_, err := iface.getSocketPath(slot)
+	return err
+}
+
+func (iface *pkcs11Interface) SecCompPermanentSlot(spec *seccomp.Specification, slot *snap.SlotInfo) error {
+	if !implicitSystemPermanentSlot(slot) {
+		spec.AddSnippet(pkcs11PermanentSlotSecComp)
+	}
+	return nil
+}
+
+func (iface *pkcs11Interface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var socketPath string
+	if err := slot.Attr("pkcs11-socket", &socketPath); err != nil || socketPath == "" {
+		return fmt.Errorf("slot %q must have a unix socket 'pkcs11-socket' attribute", slot.Ref())
+	}
+
+	// pkcs11 server unix socket
+	spec.AddSnippet(fmt.Sprintf(`# pkcs11 socket and socker env
+/{,var/}%s rw,
+# pkcs11 config for p11-proxy
+/etc/pkcs11/{,**} r,
+# pkcs11 tools
+/usr/bin/p11tool ixr,
+/usr/bin/pkcs11-tool ixr,`,
+		socketPath[1:]))
+	return nil
+}
+
+func (iface *pkcs11Interface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	// Only apply slot snippet when running as application snap
+	// on classic, slot side can be system or application
+	if !implicitSystemPermanentSlot(slot) {
```

Comment:
let's ignore any use cases on classic for now to keep this simple

Reply:
removed

Reply:
ups, that was silly copy paste :)
fixed